<launch>
    
    <param name="scan_line" type="int" value="6" />

    <!-- if 1, do mapping 10 Hz, if 2, do mapping 5 Hz. Suggest to use 1, it will adjust frequence automaticlly -->
    <param name="mapping_skip_frame" type="int" value="1" />

    <!-- remove too closed points -->
    <param name="minimum_range" type="double" value="0.3"/>


    <param name="mapping_line_resolution" type="double" value="0.2"/>
    <param name="mapping_plane_resolution" type="double" value="0.4"/>

    <group>
        <!-- <node pkg="loam_horizon" type="horizon_merger" name="horizon_merger" output="log">
            <param name="to_merge_cnt" type="int" value="2" />
            <remap from="/horizon_topic" to="/livox_pcl1" />
            <remap from="/velodyne_points" to="/merged_pcl_l0" />
        </node> -->

        <node pkg="loam_horizon" type="scanRegistration" name="scanRegistration" output="log" >
            <remap from="/livox_undistort" to="/livox_pcl0" />
        </node>
    </group>

    <group>
        <!-- <node pkg="loam_horizon" type="horizon_merger" name="horizon_merger_l1" output="log">
            <param name="to_merge_cnt" type="int" value="2" />
            <remap from="/horizon_topic" to="/livox_pcl0" />
            <remap from="/velodyne_points" to="/merged_pcl_l1" />
        </node> -->

        <node pkg="loam_horizon" type="scanRegistration" name="scanRegistration_l1" output="log">
            <remap from="/livox_undistort" to="/livox_pcl1" />
            <remap from="/velodyne_cloud_2" to="/fullres_pcl_2_l1" />
            <remap from="/laser_cloud_sharp" to="/laser_cloud_sharp_l1" />
            <remap from="/laser_cloud_less_sharp" to="/laser_cloud_less_sharp_l1" />
            <remap from="/laser_cloud_flat" to="/laser_cloud_flat_l1" />
            <remap from="/laser_cloud_less_flat" to="/laser_cloud_less_flat_l1" />
            <remap from="/laser_remove_points" to="/laser_remove_points_l1" />
            <remap from="/curvature" to="/curvature_l1" />
        </node>
    </group>

    <node pkg="loam_horizon" type="laserOdometry" name="laserOdometry" output="log" />

    <param name="/mlcalib/gather_count" type="int" value="300" />
    <param name="/mlcalib/gather_skip" type="int" value="7" />
    <rosparam param="/mlcalib/initial_extrinsic_t">[0.00,-0.13, 0.00]</rosparam>
    <rosparam param="/mlcalib/initial_extrinsic_ypr">[30.0, 0.0, 0.0]</rosparam><!-- in degrees -->

    <node pkg="loam_horizon" type="laserMapping" name="laserMapping" output="screen" >
    </node>

    <node pkg="loam_horizon" type="livox_repub" name="livox_repub" output="log" />

    <arg name="rviz" default="true" />
    <group if="$(arg rviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find loam_horizon)/rviz_cfg/mlcalib.rviz" />
    </group>

</launch>
